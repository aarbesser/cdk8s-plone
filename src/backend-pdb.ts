// eslint-disable-next-line import/no-extraneous-dependencies
import { Construct } from 'constructs';
import { IntOrString, KubePodDisruptionBudget } from './imports/k8s';

export interface PloneBackendPDBOptions {
  /**
   * maxUnavailable specification
   * @default - none
   */
  readonly maxUnavailable?: number | string;

  /**
   * minAvailable specification.
   * @default 1
   */
  readonly minAvailable?: number | string;

  /**
   * Selector label.
   */
  readonly selectorLabel: { [name: string]: string };

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };
}

export class PloneBackendPDB extends Construct {

  constructor(scope: Construct, id: string, options: PloneBackendPDBOptions) {
    super(scope, id);

    var maxUnavailable: IntOrString = IntOrString.fromString(''); // default value
    if (typeof maxUnavailable === 'number') {
      maxUnavailable = IntOrString.fromNumber(options.maxUnavailable as number);
    } else if (typeof maxUnavailable === 'string') {
      maxUnavailable = IntOrString.fromString(options.maxUnavailable as string);
    }
    var minAvailable: IntOrString = IntOrString.fromString(''); // default value
    if (typeof minAvailable === 'number') {
      minAvailable = IntOrString.fromNumber(options.minAvailable as number);
    } else if (typeof minAvailable === 'string') {
      minAvailable = IntOrString.fromString(options.minAvailable as string);
    }

    new KubePodDisruptionBudget(this, 'PDB', {
      metadata: {
        labels: options.labels ?? {},
      },
      spec: {
        selector: { matchLabels: options.selectorLabel },
        maxUnavailable: maxUnavailable,
        minAvailable: minAvailable,
      },
    });
  }
}
