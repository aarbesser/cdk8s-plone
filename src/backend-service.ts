// eslint-disable-next-line import/no-extraneous-dependencies
import { Construct } from 'constructs';
import { IntOrString, KubeServiceProps, KubeService } from './imports/k8s';

export interface PloneBackendServiceOptions {
  /**
   * Port number.
   * @default 8080
   */
  readonly port?: number;

  /**
   * Port number.
   * @default 8080;
   */
  readonly targetPort?: number;

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };
}

export class PloneBackendService extends Construct {

  constructor(scope: Construct, id: string, options: PloneBackendServiceOptions = {}) {
    super(scope, id);

    const port = options.port ?? 8080;
    const targetPort = IntOrString.fromNumber(options.targetPort ?? 8080);
    const labels = options.labels ?? {};
    const name = id + '-service';

    const serviceOpts: KubeServiceProps = {
      metadata: {
        name: name,
      },
      spec: {
        type: 'ClusterIP',
        clusterIp: 'None',
        ports: [{ port: port, targetPort: targetPort }],
        selector: labels,
      },
    };
    new KubeService(this, 'service', serviceOpts);
  }
}
