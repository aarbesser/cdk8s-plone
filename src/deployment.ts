import { Names } from 'cdk8s';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Construct } from 'constructs';
import * as k8s from './imports/k8s';
import { PlonePDB, PlonePDBOptions } from './pdb';

export interface PloneImageOptions {
  /**
   * Specify a custom image for Plone .
   * @default ""
   */
  readonly image?: string;

  /**
   * Specify Secret to pull image .
   * @default ""
   */
  readonly imagePullSecret?: string;

  /**
   * Specify Pull Policy .
   * @default ""
   */
  readonly imagePullPolicy?: string;
}

export interface PloneDeploymentOptions {
  /**
   * Specify a custom image for Plone .
   * @default "plone/plone-backend:latest"
   */
  readonly image?: PloneImageOptions;

  /**
 * Number of replicas.
 * @default 2
 */
  readonly replicas?: number;

  /**
   * Port number.
   */
  readonly port: number;

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };

  /**
   * Extra container spec to use for ploencontainer.
   * @default - []
   */
  readonly ploneContainer?: k8s.Container;

  /**
   * Sidecar container spec to associate with resources.
   * @default - []
   */
  readonly sidecarContainers?: k8s.Container[];

  /**
   * Create a PodDisruptionBugdet for the deployment?
   * If given
   * @default - none
   */
  readonly pdb?: PlonePDBOptions;
}

export class PloneDeployment extends Construct {

  constructor(scope: Construct, id: string, options: PloneDeploymentOptions) {
    super(scope, id);
    const image = options.image ?? {};
    const replicas = options.replicas ?? 2;
    const label = { app: Names.toLabelValue(this) };
    const template_labels = {
      ...options.labels ?? {},
      ...label,
    };
    var ploneContainerSpec = {
      ...options.ploneContainer ?? {},
      name: id + '-container', // here the namespaced name shold be used, but how?
      image: image.image,
      imagePullPolicy: image.imagePullPolicy,
      imagePullSecret: { name: image.imagePullSecret },
    };
    const deploymentOptions: k8s.KubeDeploymentProps = {
      metadata: {
        labels: options.labels ?? {},
      },
      spec: {
        replicas,
        selector: {
          matchLabels: label,
        },
        template: {
          metadata: { labels: template_labels },
          spec: {
            containers: [
              ploneContainerSpec,
              ...options.sidecarContainers ?? [],
            ],
          },
        },
      },
    };

    new k8s.KubeDeployment(this, 'deployment', deploymentOptions);

    if (options.pdb ?? false) {
      const pdbOptions = options.pdb ?? {};
      new PlonePDB(this, 'pdb', label, pdbOptions);
    }
  }
}
