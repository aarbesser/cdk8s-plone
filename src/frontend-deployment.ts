import { Names } from 'cdk8s';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Construct } from 'constructs';
import * as k8s from './imports/k8s';
import { PlonePDB, PlonePDBOptions } from './pdb';

export interface PloneFrontendDeploymentOptions {
  /**
   * Specify a custom image for Plone Frontend.
   * @default "plone/plone-frontend:latest"
   */
  readonly image?: string;

  /**
 * Number of replicas.
 * @default 2
 */
  readonly replicas?: number;

  /**
   * Port number.
   * @default 8080
   */
  readonly port?: number;

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };

  /**
   * Create a PodDisruptionBugdet for the deployment?
   * If given
   * @default - none
   */
  readonly pdbOptions?: PlonePDBOptions;
}

export class PloneFrontendDeployment extends Construct {

  constructor(scope: Construct, id: string, options: PloneFrontendDeploymentOptions = {}) {
    super(scope, id);
    const image = options.image ?? 'plone/plone-backend:latest';
    const replicas = options.replicas ?? 2;
    const label = { app: Names.toLabelValue(this) };
    const template_labels = {
      ...options.labels ?? {},
      ...label,
    };
    const deploymentOptions: k8s.KubeDeploymentProps = {
      metadata: {
        labels: options.labels ?? {},
      },
      spec: {
        replicas,
        selector: {
          matchLabels: label,
        },
        template: {
          metadata: { labels: template_labels },
          spec: {
            containers: [
              {
                name: id + '-container', // here the namespaced name shold be used, but how?
                image: image,
              },
            ],
          },
        },
      },
    };

    new k8s.KubeDeployment(this, 'deployment', deploymentOptions);


    if (options.pdbOptions ?? false) {
      const pdbOptions = options.pdbOptions ?? {};
      new PlonePDB(this, id + '-pdb', label, pdbOptions);
    }
  }
}
