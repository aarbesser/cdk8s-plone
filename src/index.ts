export { Plone, PloneOptions, PloneBaseOptions } from './plone';
