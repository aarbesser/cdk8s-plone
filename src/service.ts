// eslint-disable-next-line import/no-extraneous-dependencies
import { Construct } from 'constructs';
import { IntOrString, KubeServiceProps, KubeService } from './imports/k8s';

export interface PloneServiceOptions {

  /**
   * targetPort number.
   */
  readonly targetPort: number;

  /**
   * Selector label.
   */
  readonly selectorLabel: { [name: string]: string };

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };
}

export class PloneService extends Construct {

  constructor(scope: Construct, id: string, options: PloneServiceOptions) {
    super(scope, id);

    const targetPort = IntOrString.fromNumber(options.targetPort);
    const selectorLabel = options.selectorLabel;

    const serviceOpts: KubeServiceProps = {
      metadata: {
        labels: options.labels ?? {},
      },
      spec: {
        type: 'ClusterIP',
        clusterIp: 'None',
        ports: [{ port: 80, targetPort: targetPort }],
        selector: selectorLabel,
      },
    };
    new KubeService(this, 'service', serviceOpts);
  }
}
