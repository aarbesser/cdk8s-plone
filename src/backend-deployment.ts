import { Names } from 'cdk8s';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Construct } from 'constructs';
import { PloneBackendPDB, PloneBackendPDBOptions } from './backend-pdb';
import * as k8s from './imports/k8s';

export interface PloneBackendDeploymentOptions {
  /**
   * Specify a custom image for Plone Backend.
   * @default "plone/plone-backend:latest"
   */
  readonly image?: string;

  /**
 * Number of replicas.
 * @default 2
 */
  readonly replicas?: number;

  /**
   * Port number.
   * @default 8080
   */
  readonly port?: number;

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };

  /**
   * Create a PodDisruptionBugdet for the deployment?
   * If given
   * @default - none
   */
  readonly pdbOptions?: PloneBackendPDBOptions;
}

export class PloneBackendDeployment extends Construct {

  constructor(scope: Construct, id: string, options: PloneBackendDeploymentOptions = { }) {
    super(scope, id);
    const image = options.image ?? 'plone/plone-backend:latest';
    const replicas = options.replicas ?? 2;
    const label = { app: Names.toLabelValue(this) };
    const template_labels = {
      ...options.labels ?? {},
      ...label,
    };
    const pdb = options.pdbOptions ?? true;
    const deploymentOptions: k8s.KubeDeploymentProps = {
      metadata: {
        labels: options.labels ?? {},
      },
      spec: {
        replicas,
        selector: {
          matchLabels: label,
        },
        template: {
          metadata: { labels: template_labels },
          spec: {
            containers: [
              {
                name: id + '-container', // here the namespaced name shold be used, but how?
                image: image,
              },
            ],
          },
        },
      },
    };

    new k8s.KubeDeployment(this, 'deployment', deploymentOptions);

    if (pdb ?? false) {
      const pdbOptions = {
        ...options.pdbOptions ?? {},
        selector_label: { app: Names.toLabelValue(this) },
      };
      new PloneBackendPDB(this, 'pdb', pdbOptions);
    }
  }
}
