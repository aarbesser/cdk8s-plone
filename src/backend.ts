// eslint-disable-next-line import/no-extraneous-dependencies
import { Construct } from 'constructs';
import { PloneBackendDeploymentOptions, PloneBackendDeployment } from './backend-deployment';
import { PloneBackendServiceOptions, PloneBackendService } from './backend-service';

export interface PloneBackendOptions {
  readonly deployment?: PloneBackendDeploymentOptions;
  readonly service?: PloneBackendServiceOptions;
}

export class PloneBackend extends Construct {

  constructor(scope: Construct, id: string, options: PloneBackendOptions = {}) {
    super(scope, id);
    const deployment = options.deployment ?? {};
    const service = options.service ?? {};

    // Create a deployment
    new PloneBackendDeployment(this, 'deployment', deployment);

    // Create a service
    new PloneBackendService(this, 'service', service);
  }
}
