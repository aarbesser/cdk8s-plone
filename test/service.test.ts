import { Chart, Testing } from 'cdk8s';
import { PloneService } from '../src/service';


test('defaults', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'plone');

  // WHEN
  new PloneService(chart, 'test', { targetPort: 8080, selectorLabel: { app: 'plone' } });

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});

