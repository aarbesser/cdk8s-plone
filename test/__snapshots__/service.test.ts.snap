// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults 1`] = `
[
  {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": {
      "name": "plone-test-service-c8a1159e",
    },
    "spec": {
      "clusterIP": "None",
      "ports": [
        {
          "port": 80,
          "targetPort": 8080,
        },
      ],
      "selector": {
        "app": "plone",
      },
      "type": "ClusterIP",
    },
  },
]
`;
